apiserver/facades/client/highavailability/highavailability.go

```diff
@@ -148,6 +150,10 @@ func (api *HighAvailabilityAPI) enableHASingle(st *state.State, spec params.Cont
 	}
 	spec.Constraints.Spaces = cfg.AsSpaceConstraints(spec.Constraints.Spaces)
 
+	if err = validatePlacementForSpaces(st, spec.Constraints.Spaces, spec.Placement); err != nil {
+		return params.ControllersChanges{}, errors.Trace(err)
+	}
+
 	// Might be nicer to pass the spec itself to this method.
 	changes, err := st.EnableHA(spec.NumControllers, spec.Constraints, spec.Series, spec.Placement, api.machineID)
```

Comment:
I'm wondering if we should be doing the check in APIServer, or whether this light about "validatePlacementForSpaces" should actually be taking place inside "st.EnableHA".
We have spec.Constraints and spec.Placement available inside st.EnableHA, and it feels appropriate for the internal State workings to disallow placement regardless of whether the request was via the API or not (while being via the API is the only way it can actually be triggered.)

I'm not settled on it, just trying to think through what layer has the responsibility of enforcing that the target machines do have the right spaces.

I suppose I should have asked the same question on 'validateCurrentControllers", and I do like the idea of having the validation steps next to each other..
Mabye this is more something to do in a follow up? (and/or maybe I'm wrong about what layer should be enforcing it)

Reply:
I can go either way on this.

On one hand pushing it into state is more full-proof regarding possible error scenarios, as you say; it guarantees more correctness.

On the other, I feel like _APIServer_ is synonymous with (direct, non-HTTP) API if you set aside facades, and feels like the correct layer to do this kind of validation. 

I know _state_ isn't "pure" in that there are plenty of places where it has more than just the logic required to store our representations, but something about pushing more logic into it gives me pause.

One justification for it might be working with data in a more efficient way than the entity level - for-purpose queries like you did to replace _VotingMemberIds_. This makes such state operations more "PL/SQL-ish". For (rough) example, on a hot path it might well be better to pull out addresses for all the machines you care about at once rather than decoding into machine entities one by one in a loop.

Reply:
Yeah, I think its not a big deal at this point. Let's leave it in the apiserver layer for now.